### PR TITLE
feat: kong provider implementation

### LINKED ISSUE #63: Add Kong provider
**What would you like to be added**:

We'd like to add a new provider to convert ingress resources in setups that use the [Kong Kubernetes Ingress Controller](https://github.com/Kong/kubernetes-ingress-controller). 

**Why this is needed**:

This is needed to give the Kong Kubernetes Ingress Controller users a tool to migrate from Ingress API to Gateway API.

### PR DESCRIPTION
**Pull request by mlavacca**
**What type of PR is this?**

/kind feature

**What this PR does / why we need it**:

Implementation of the Kong provider with a basic set of additional features:
- method matching
- headers matching

**Which issue(s) this PR fixes**:

Fixes #63 

**Does this PR introduce a user-facing change?**:

```release-note
Kong provider addition. Now, a user can specify "kong" as the provider and leverage the konghq.com/headers and konghq.com/methods features of the Kong Ingress Controller.
```

### CODE REVIEW

**Comment by levikobi** [comment]
Thanks for jumping in to do this one @mlavacca 🥳

I did a quick first review, and so far, it looks like a great contribution to me. I will try to go through it more deeply as soon as possible.

Do you mind adding a README.md file to the `ingresskong` package explaining the annotations you added support for (similarly to ingressnginx's README.md)?

Also, if you have any suggestions regarding the project layout, I'd love to hear them. If you had any struggles that you think we can improve for future provider implementors.

Thanks!

**Comment by mlavacca** [comment]
> Thanks for jumping in to do this one @mlavacca 🥳
> 
> I did a quick first review, and so far, it looks like a great contribution to me. I will try to go through it more deeply as soon as possible.

Thanks for reviewing it! :)

> 
> Do you mind adding a README.md file to the `ingresskong` package explaining the annotations you added support for (similarly to ingressnginx's README.md)?
> 

Sure, I'll do it 👍 

> Also, if you have any suggestions regarding the project layout, I'd love to hear them. If you had any struggles that you think we can improve for future provider implementors.
> 

There are a couple of things that promptly come to my mind:
1. How providers are named. The Nginx provider is named `ingress-nginx`, and if we follow the same pattern, Kong should be called `ingress-kong`. Given that this is a tool to help migrate from ingress, having all the implementations named with the `ingress-` prefix sounds a bit redundant to me. Maybe just `nginx`, `kong`, `whatever`.
2. Where the feature files are put. So far, the features are put in the `providers/<implementation_name>` folder, along with the main files (`resource_reader.go`, `converter.go`, ...). I think it would be beneficial to create a sub-directory under `providers/<implementation_name>` with all the features provided by that implementation. By doing so, we can just export the `<feature_name>Feature` functions and leave all the helpers used by such a feature unexported. I think that by following this pattern, the maintainability of a provider can be kept under control, even with the increase of the supported features.

File: pkg/i2gw/providers/ingresskong/README.md [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+- `konghq.com/methods`: If specified, the values of this annotation are used to
+  perform method matching on the associated ingress rules. Multiple methods can
+  be specified by separating values with commas. Example: `konghq.com/methods: "POST,GET"`.
+- `konghq.com/headers.*`: If specified, the values of this annotation are used to
+  perform header matching on the associated ingress rules. The header name is specified
+  in the annotation key after `.`, and the annotations value can contain multiple
+  header values separated by commas. All the header values for a specific header
+  name are intended to be ORed. Example: `konghq.com/headers.x-routing: "alpha,bravo"`.
+
+If you are reliant on any annotations not listed above, please open an issue.
```
Comment (levikobi) [inline pkg/i2gw/providers/ingresskong/README.md:16]: Awesome! Can you please add a link to it at https://github.com/kubernetes-sigs/ingress2gateway#current-supported-providers ?

File: pkg/i2gw/providers/ingresskong/converter.go [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+		featureParsers: []i2gw.FeatureParser{
+			headerMatchingFeature,
+			methodMatchingFeature,
+		},
+	}
+}
+
+// ToGatewayAPIResources converts the received i2gw.InputResources to i2gw.GatewayResources
+// including the ingress-kong specific features.
+func (c *converter) ToGatewayAPI(resources i2gw.InputResources) (i2gw.GatewayResources, field.ErrorList) {
```
Comment (levikobi) [inline pkg/i2gw/providers/ingresskong/converter.go:45]: Can you please rename the documentation to `ToGatewayAPI`? (I know ingress-nginx implementation is also inconsistent, this function was renamed multiple times, so I missed it at the time)

**Review by levikobi — COMMENTED** [review]
Great work, thanks again @mlavacca!
/lgtm

**Comment by levikobi** [comment]
> There are a couple of things that promptly come to my mind:
> 
> 1. How providers are named. The Nginx provider is named `ingress-nginx`, and if we follow the same pattern, Kong should be called `ingress-kong`. Given that this is a tool to help migrate from ingress, having all the implementations named with the `ingress-` prefix sounds a bit redundant to me. Maybe just `nginx`, `kong`, `whatever`.
> 2. Where the feature files are put. So far, the features are put in the `providers/<implementation_name>` folder, along with the main files (`resource_reader.go`, `converter.go`, ...). I think it would be beneficial to create a sub-directory under `providers/<implementation_name>` with all the features provided by that implementation. By doing so, we can just export the `<feature_name>Feature` functions and leave all the helpers used by such a feature unexported. I think that by following this pattern, the maintainability of a provider can be kept under control, even with the increase of the supported features.

Yea, I can agree with both points honestly. I guess we can open a PR for that at some point

File: pkg/i2gw/providers/ingresskong/converter.go (hunk 2 above)
Comment (mlavacca) [reply pkg/i2gw/providers/ingresskong/converter.go]: Yep, I've renamed both the kong and the nginx one.

File: pkg/i2gw/providers/ingresskong/README.md (hunk 1 above)
Comment (mlavacca) [reply pkg/i2gw/providers/ingresskong/README.md]: 👍 added

**Review by levikobi — APPROVED** [review]
/lgtm

**Review by robscott — COMMENTED** [review]
Thanks for all the work on this @mlavacca!

/approve

**Pull request merged**